tensorflow_lite_support/ios/task/processor/sources/TFLSegmentationResult.m

```diff
@@ -13,86 +13,99 @@
 }
 
-- (id)copyWithZone:(NSZone *)zone {
-  return [[TFLCategoryMask alloc] initWithWidth:self.width
-                                         height:self.height
-                                           mask:self.mask];
-}
-
 - (void)dealloc {
   free(self.mask);
 }
 
 @end
 
-@implementation TFLConfidenceMask {
-  NSInteger _width;
-  NSInteger _height;
-  float *_mask;
-}
+@implementation TFLConfidenceMask
 
 - (instancetype)initWithWidth:(NSInteger)width height:(NSInteger)height mask:(float *)mask {
   self = [super init];
   if (self) {
     _width = width;
     _height = height;
     if (mask != NULL) {
-      _mask = malloc(width * height * sizeof(float));
-      memcpy(_mask, mask, width * height * sizeof(float));
+      _mask = [TFLCommonUtils mallocWithSize:width * height * sizeof(float) error:nil];
+      if (_mask) {
+        memcpy(_mask, mask, width * height * sizeof(float));
+      }
     }
   }
   return self;
 }
 
-- (id)copyWithZone:(NSZone *)zone {
-  return [[TFLConfidenceMask alloc] initWithWidth:self.width
-                                           height:self.height
-                                             mask:self.mask];
-}
-
 - (void)dealloc {
   free(self.mask);
 }
 
 @end
 
 @implementation TFLColoredLabel
-@synthesize r;
-@synthesize g;
-@synthesize b;
-@synthesize label;
-@synthesize displayName;
+
+- (instancetype)initWithRed:(NSUInteger)r green:(NSUInteger)g blue:(NSUInteger)b label:(NSString *)label displayName:(NSString *)displayName {
+  self = [super init];
+  if (self) {
+    _r = r;
+    _g = g;
+    _b = b;
+    _label = [label copy];
+    _displayName = [displayName copy];
+  }
+  return self;
+}
 
 @end
 
 @implementation TFLSegmentation
-@synthesize confidenceMasks;
-@synthesize categoryMask;
-@synthesize coloredLabels;
+
+- (instancetype)initWithConfidenceMasks:(NSArray<TFLConfidenceMask *> *)confidenceMasks  coloredLabels:(NSArray<TFLColoredLabel *> *)coloredLabels {
+  return [self initWithConfidenceMasks:confidenceMasks categoryMask:nil coloredLabels:coloredLabels];
+}
+
+- (instancetype)initWithCategoryMask:(TFLCategoryMask *)categoryMask coloredLabels:(NSArray<TFLColoredLabel *> *)coloredLabels {
+  return [self initWithConfidenceMasks:nil categoryMask:categoryMask coloredLabels:coloredLabels];
+}
+
+- (instancetype)initWithConfidenceMasks:(NSArray<TFLConfidenceMask *> *)confidenceMasks categoryMask:(TFLCategoryMask *)categoryMask coloredLabels:(NSArray<TFLColoredLabel *> *)coloredLabels {
+  self = [super init];
+  if (self) {
+    _confidenceMasks = [confidenceMasks copy];
+    _categoryMask = categoryMask; // Not copying since not an array and all properties are readonly.
+    _coloredLabels = [coloredLabels copy];
+  }
+  return self;
+}
 
 @end
 
 @implementation TFLSegmentationResult
-@synthesize segmentations;
+- (instancetype)initWithSegmentations:(NSArray<TFLSegmentation *> *)segmentations {
+  self = [super init];
+  if (self) {
+    _segmentations = [segmentations copy];
```

Comment:
Let's avoid excessive copy here. The only place where we have to create a copy is when converting from C object to ObjC object (e.g. (float *)mask to TFLConfidenceMask) because we need to manage the ownership of the object. Otherwise, just keeping a strong reference should be enough because the data type used is immutable.

Reply:
Fixed. Reverted copying for all strings and arrays.